src/repr/sdd.rs

```diff
@@ -211,54 +211,42 @@ impl SddPtr {
-        }
-        if self.is_or() {
-            let r = self.node_ref_mut();
-            if r.semantic_hash.is_some() {
-                let v = FiniteField::new(r.semantic_hash.unwrap());
-                return if self.is_neg() { v.negate() } else { v };
+        match self {
+            PtrTrue => FiniteField::new(1),
+            PtrFalse => FiniteField::new(0),
+            Var(v, b) => {
+                let (l_w, h_w) = map.get_var_weight(*v);
+                return if *b { *h_w } else { *l_w };
             }
-            // no cached value, compute it
-            let h = self.node_iter().fold(FiniteField::new(0), |acc, i| {
-                acc + i.prime().cached_semantic_hash(vtree, map)
-                    * i.sub().cached_semantic_hash(vtree, map)
-            });
-            self.node_ref_mut().semantic_hash = Some(h.value());
-            if self.is_neg() {
-                return h.negate();
-            } else {
-                return h;
-            };
-        }
-        if self.is_var() {
-            let v = self.get_var();
-            let (h_w, l_w) = map.get_var_weight(v.get_label());
-            if v.get_polarity() {
-                return *h_w;
-            } else {
-                return *l_w;
+            BDD(_) => {
+                let b = self.mut_bdd_ref();
+                if b.semantic_hash.is_some() {
+                    return FiniteField::new(b.semantic_hash.unwrap());
+                }
+
+                // no cached value, compute it
+                let l_h = self.low().cached_semantic_hash(vtree, map);
+                let h_h = self.high().cached_semantic_hash(vtree, map);
+                let (l_w, h_w) = map.get_var_weight(self.topvar());
+                let h = (*l_w) * l_h + (*h_w) * h_h;
+                b.semantic_hash = Some(h.value());
+                h
             }
-        }
-        if self.is_bdd() {
-            let b = self.mut_bdd_ref();
-            if b.semantic_hash.is_some() {
-                let h = FiniteField::new(b.semantic_hash.unwrap());
-                return if self.is_neg() { h.negate() } else { h };
+            Reg(_) => {
+                let r = self.node_ref_mut();
+                if r.semantic_hash.is_some() {
+                    return FiniteField::new(r.semantic_hash.unwrap());
+                }
+
+                // no cached value, compute it
+                let h = self.node_iter().fold(FiniteField::new(0), |acc, i| {
+                    acc + i.prime().cached_semantic_hash(vtree, map)
+                        * i.sub().cached_semantic_hash(vtree, map)
+                });
+                self.node_ref_mut().semantic_hash = Some(h.value());
+                h
             }
-            let var = self.topvar();
-            let l_h = self.low_raw().cached_semantic_hash(vtree, map);
-            let h_h = self.high_raw().cached_semantic_hash(vtree, map);
-            let (h_w, l_w) = map.get_var_weight(var);
```

Comment:
In every other place in the code, this is `l_w`, `h_w`. Flipped it - but that didn't resolve the problem.